src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java

```diff
@@ -0,0 +1,218 @@
+  private final Region region;
+  private final SecretsManagerClient secretsManagerClient;
+  private final GetSecretValueRequest getSecretValueRequest;
+
+  public AWSSecretsManagerPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) throws SQLException {
+
+    this(
+        currentConnectionProvider,
+        propertySet,
+        nextPlugin,
+        logger,
+        null,
+        null
+    );
+  }
+
+  AWSSecretsManagerPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger,
+      SecretsManagerClient secretsManagerClient,
+      GetSecretValueRequest getSecretValueRequest) throws SQLException {
+
+    try {
+      Class.forName("software.amazon.awssdk.services.secretsmanager.SecretsManagerClient");
+    } catch (ClassNotFoundException e) {
+      logger.logError(ERROR_MISSING_DEPENDENCY);
+      throw new SQLException(ERROR_MISSING_DEPENDENCY);
+    }
+
+    this.nextPlugin = nextPlugin;
+    this.logger = logger;
+    this.secretId = propertySet.getStringProperty(SECRET_ID_PROPERTY).getValue();
+    this.region = Region.of(propertySet.getStringProperty(REGION_PROPERTY).getValue());
+
+    if (secretsManagerClient != null && getSecretValueRequest != null) {
+      this.secretsManagerClient = secretsManagerClient;
+      this.getSecretValueRequest = getSecretValueRequest;
+
+    } else {
+      this.secretsManagerClient = SecretsManagerClient.builder()
+          .region(this.region)
+          .build();
+      this.getSecretValueRequest = GetSecretValueRequest.builder()
+          .secretId(this.secretId)
+          .build();
+    }
+  }
+
+  @Override
+  public void openInitialConnection(ConnectionUrl connectionUrl) throws SQLException {
+    final Properties updatedProperties = new Properties();
+    updatedProperties.putAll(connectionUrl.getOriginalProperties());
+    final Pair<String, Region> secretKey = Pair.of(secretId, region);
+    Secret secret = SECRET_CACHE.get(secretKey);
+    ConnectionUrl newConnectionUrl;
+
+    try {
+      // Attempt to open initial connection with cached secret
+      updateConnectionProperties(updatedProperties, secret);
+      newConnectionUrl = ConnectionUrl.getConnectionUrlInstance(connectionUrl.getDatabaseUrl(), updatedProperties);
+      this.nextPlugin.openInitialConnection(newConnectionUrl);
+
+    } catch (SQLException connectionFailedException) {
+      // If the thrown error is because user access was denied, try again with new credentials
+      if (SQLSTATE_ACCESS_ERROR.equals(connectionFailedException.getSQLState())) {
+        try {
+          // Get updated credentials and store the secret in the cache
+          secret = getCurrentCredentials();
+          SECRET_CACHE.put(secretKey, secret);
+          updateConnectionProperties(updatedProperties, secret);
```

Comment:
These two lines should be after the catch clause, right above `newConnectionUrl = ...`